torch/lib/c10d/ProcessGroupNCCL.hpp

```diff
@@ -120,28 +120,39 @@ class ProcessGroupNCCL : public ProcessGroup {
     friend class ProcessGroupNCCL;
   };
 
-  // Constructor will also check the number of available GPUs in the system
+  // If you wish to create multiple process groups, each with a potentially
+  // different rank and size, you can do so by passing a new store instance
+  // to each one. If you have only a single store object, you can
+  // use the `c10d::PrefixStore` to derive scoped instances.
+  // This is also what the Python API in torch.distributed does.
   //
-  // Group support:
+  // The process group instance keeps a reference to the store because
+  // it may be used long after the constructor runs. In fact, the constructor
+  // doesn't create any NCCL communicators. A single NCCL communicator can
+  // only be used on a specific set of devices, and are therefore created
+  // on-demand when a collective runs. If another collective is executed later,
+  // against a different set of devices, the process group creates another NCCL
+  // communicator. These NCCL communicators are cached and reused if possible.
   //
-  // In order to support multiple NCCL process groups, each of which has
-  // different group ranks, we need to use groupName to identify each group
-  // to ensure the correct behavior. In other words, each process group that
-  // has different group ranks needs to have a different and unique groupName
-  // to avoid clashing into undefined behaviors.
-  //
-  // In Python frontend API of torch.distributed, it guarantees that each group
-  // will have a unique name to be passed into the ProcessGroupNCCL constructor.
-  // If you would like to use ProcessGroupNCCL constructor directly, it is
-  // your reponsibility to do so as well.
   ProcessGroupNCCL(
       const std::shared_ptr<Store>& store,
       int rank,
       int size,
-      const std::string& groupName = "",
       const std::chrono::milliseconds& opTimeout =
           std::chrono::milliseconds(kProcessGroupNCCLOpTimeoutMillis));
 
+  // This constructor includes the deprecated `groupName` argument.
+  // If you have existing code that uses the `groupName`, you can replace
+  // it by specifying a `c10d::PrefixStore(groupName, store)` for store.
+  C10_DEPRECATED ProcessGroupNCCL(
```

Comment:
Where is this constructor used? We are not exposing this to Python anyway, why do we need to keep a constructor with `groupName`?

Reply:
This is to not break downstream C++ users. I think torchcraft uses this directly.